src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -375,127 +330,206 @@ let ApplyAllOptimizations
-                    phase2And3Settings,
-                    ccu,
-                    tcGlobals,
-                    tcVal,
-                    importMap,
-                    env,
-                    isIncrementalFragment,
-                    tcConfig.fsiMultiAssemblyEmit,
-                    tcConfig.emitTailcalls,
-                    hidden,
-                    implFile
-                )
-
-            optEnvExtraLoop, implFile
-        else
-            env, implFile
-
-    let phase3 (env: Optimizer.IncrementalOptimizationEnv, hidden: SignatureHidingInfo, implFile: CheckedImplFile) =
-        use _ =
-            FSharp.Compiler.Diagnostics.Activity.start "phase3" [| "QualifiedNameOfFile", implFile.QualifiedNameOfFile.Text |]
-
-        let implFile =
-            if tcConfig.doDetuple then
-                let implFile = implFile |> Detuple.DetupleImplFile ccu tcGlobals
-                implFile
-            else
-                implFile
-
-        let implFile =
-            if tcConfig.doTLR then
-                implFile
-                |> InnerLambdasToTopLevelFuncs.MakeTopLevelRepresentationDecisions ccu tcGlobals
-            else
-                implFile
-
-        let implFile = LowerCalls.LowerImplFile tcGlobals implFile
-
-        if tcConfig.doFinalSimplify then
-            let (optEnvFinalSimplify, implFile, _, _), _ =
-                Optimizer.OptimizeImplFile(
-                    phase2And3Settings,
-                    ccu,
-                    tcGlobals,
-                    tcVal,
-                    importMap,
-                    env,
-                    isIncrementalFragment,
-                    tcConfig.fsiMultiAssemblyEmit,
-                    tcConfig.emitTailcalls,
-                    hidden,
-                    implFile
-                )
-
-            optEnvFinalSimplify, implFile
-        else
-            env, implFile
+    let measurements = ConcurrentDictionary<string, int64>()
 
-    let results, optEnvFirstLoop =
-        match tcConfig.optSettings.processingMode with
-        | Optimizer.OptimizationProcessingMode.PartiallyParallel ->
-            let ct = CancellationToken.None
+    let measure (name: string) =
+        let sw = Stopwatch.StartNew()
 
-            let results, optEnvFirstPhase =
-                ParallelOptimization.optimizeFilesInParallel optEnv (phase1, phase2, phase3) implFiles ct
+        { new System.IDisposable with
+            member this.Dispose() =
+                lock measurements (fun () ->
+                    if measurements.ContainsKey name = false then
+                        measurements[name] <- 0L
 
-            results |> Array.toList, optEnvFirstPhase
-        | Optimizer.OptimizationProcessingMode.Sequential ->
-            let results, (optEnvFirstLoop, _, _, _) =
-                ((optEnv, optEnv, optEnv, SignatureHidingInfo.Empty), implFiles)
+                    measurements[name] <- measurements[name] + sw.ElapsedMilliseconds)
```

Comment:
Since using ConcurrenctDictionairy, I would recommend to turn it into an `ConcurrentDictionary<string, Ref<int64>>() `and then incrementing like this:

```
let refCell = measurements.GetOrAdd(name,ref 0L)
System.Threading.Interlocked.Add(refCell,sw.ElapsedMilliseconds) |> ignore
```

Reply:
I'll remove this measuring code.